Documents/AnalyticsToolDesignDocument.md

```diff
@@ -0,0 +1,108 @@
+- The **Git index** is git's internal list of files which it is tracking for creating  the next `commit`. It is useful to this project by means of comparing it to the list of **placeholders** to deduce if those files should truly be hydrated or if it may have occurred incidentally.
+- The **file system** must be carefully considered in the case of working with a mounted repository. VFS4G intercepts file I/O calls at the kernel level to inject its own behavior (the projected file system and if needed a network call to pull down the file which the system asked for in the correct format). Because of this, enumerating directories can cause additional data to be pulled down from the remote by hydrating extra files or further hydrating those that already exist on disk. Accordingly, care must be taken when interacting with the file system lest the tool designed to help users increase enlistment heath end up causing more problems.
+
+## Components
+
+- **API** - The server component of the client-server relationship. Responsible for the logic and computation of statistical data, as well as the assembly and sending of a **json statistics object** over the named pipe to the client. It sits within the `InProgressMount` process and interacts with the VFS4G internals
+- **Verb** - The `gvfs statistics` verb will be the component that handles the client side interaction with the API. When run, it will collect and store the statistics from the enlistment to either be displayed immediately or collected for trend purposes to then provide understanding in the future. This is also the point where data can be sent to telemetry for our internal analysis. The verb will also output the data to the console with a tool to help visualize and understand the data. It will output the analytics data to the command line and will also be used during the design period to test the API and verb
+- **Utility tools for dehydration** - *Stretch goal* | After the analytics are collected, these components will provide utilities to help the user get their enlistment back to a healthier state. The ideal goal is to have VFS4G suggest uses to the user (EG: `This directory was hydrated by Sublime text editor. You can dehydrate it by running "gvfs dehydrate --confirm -d /path/to/repo"`)
+- **Additional visualization tools** - *Stretch goal* | With a comprehensive API in place, the possibility opens up for different visualization tools to be created that consume the json data and display it to the user in different ways, potentially even eventually interfacing with the utility tools for an interactive GUI
+

```

Comment:
One more stretch goal to consider adding to the bottom of your list:

*Raising notifications to the user* - Is there someway we could alert the user that there's a potential issue and encourage them to run `gvfs statistics`?  Assuming the verb is non-intrusive one option might be having `GVFS.Service` run the command shortly after automatically mounting repos and presenting the user with a notification if hydration levels are over a certain threshold.

Reply:
That's a really go idea, I definitely like that a lot. I'm still trying to hash out what percentage exactly indicates an overhydrated directory. I'm guessing either somebody has a pretty good guess or we can use some data to determine that